File: backend/app/api/account.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Security
from sqlalchemy import select
from sqlalchemy.orm import Session

from ..db.dependencie import get_db
from ..models.models import Account as AccountModel
from ..schemas.schemas import AccountIn, AccountOut, Scopes, UserOut
from .oauth import get_current_user

router = APIRouter(prefix="/accounts", tags=["account"])


@router.post("/", response_model=AccountOut, status_code=201)
def create_account(
    account: AccountIn,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[UserOut, Security(get_current_user, scopes=[Scopes.USER.value])],
):
    """
    Create a new account.
    """
    account_db = db.scalar(
        select(AccountModel).where(AccountModel.name == account.name, AccountModel.user_id == current_user.id)
    )
    if account_db:
        raise HTTPException(status_code=409, detail="Account is already exists")

    account = AccountModel(**account.model_dump(), user_id=current_user.id)
    db.add(account)
    db.commit()
    db.refresh(account)
    return account


@router.get("/", response_model=list[AccountOut])
def get_accounts(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[UserOut, Security(get_current_user, scopes=[Scopes.USER.value])],
):
    """
    Gets data from all existing accounts.
    """
    accounts = db.scalars(select(AccountModel).where(AccountModel.user_id == current_user.id)).all()
    if not accounts:
        raise HTTPException(status_code=404, detail="Accounts not found")
    return accounts


@router.get("/{account_id}", response_model=AccountOut)
def get_account(
    account_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[UserOut, Security(get_current_user, scopes=[Scopes.USER.value])],
):
    """
    Gets the data of a specific account.
    """
    account = db.scalar(
        select(AccountModel).where(AccountModel.id == account_id, AccountModel.user_id == current_user.id)
    )
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    return account


@router.put("/{account_id}", response_model=AccountOut)
def update_account(
    update_data_account: AccountIn,
    account_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[UserOut, Security(get_current_user, scopes=[Scopes.USER.value])],
):
    """
    Update the data of a specific account.
    """
    account = db.scalar(
        select(AccountModel).where(AccountModel.id == account_id, AccountModel.user_id == current_user.id)
    )
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    if update_data_account.name is not None:
        account.name = update_data_account.name
    if update_data_account.currency is not None:
        account.currency = update_data_account.currency
    if update_data_account.balance is not None:
        account.balance = update_data_account.balance
    db.commit()
    return account


@router.delete("/{account_id}", status_code=204)
def delete_account(
    account_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[UserOut, Security(get_current_user, scopes=[Scopes.USER.value])],
):
    """
    Delete a specific account.
    """
    account = db.scalar(
        select(AccountModel).where(AccountModel.id == account_id, AccountModel.user_id == current_user.id)
    )
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    db.delete(account)
    db.commit()

File: frontend/web/app/src/api/account.ts
import { AccountIn, AccountOut } from "../schemas/account";
import { ErrorResponse  } from "@/schemas/error";
import { AxiosError } from 'axios';
import { axi } from "./axiosConfig";

// Get all accounts.
export const getAccounts = (token: string): Promise<AccountOut[] | ErrorResponse> => {
    return axi.get("/accounts", { headers: { 'Authorization': `Bearer ${token}` } })
    .then(response => {
        console.log(response.status);
        return response.data;
    })
    .catch((error: AxiosError) => {
        const status = error.response?.status || 500;
        const msg = error.message;

        return {
            status,
            msg
        };
    });
};

// Get a account
export const getAccount = (token: string, id: number): Promise<AccountOut | ErrorResponse> => {
    return axi.get(`/accounts/${id}`, { headers: { 'Authorization': `Bearer ${token}` } })
    .then(response => {
        console.log(response.status);
        return response.data;
    })
    .catch((error: AxiosError) => {
        const status = error.response?.status || 500;
        const msg = error.message;

        return {
            status,
            msg
        };
    });
};

// Create an account
export const createAccount = (token: string, account: AccountIn): Promise<AccountOut> => {
    return axi.post('/accounts', {
        "name": account.name,
        "currency": account.currency,
        "balance": account.balance,
    }, { headers: { 'Authorization': `Bearer ${token}` } })
    .then(response => {
        console.log(response.status);
        return response.data;
    })
    .catch((error: AxiosError) => {
        const detail: {detail: string} = error.response?.data as {detail: string};
        throw new Error(`${detail.detail}`)
    });
};

// Update an account
export const updateAccount = (token: string, id: number, account: AccountIn): Promise<AccountOut> => {
    console.log(id)
    return axi.put(`/accounts/${id}`, {
        "name": account?.name,
        "currency": account?.currency,
        "balance": account?.balance,
    }, { headers: { 'Authorization': `Bearer ${token}` } })
    .then(response => {
        console.log(response.status);
        return response.data;
    })
    .catch((error: AxiosError) => {
        const detail: {detail: string} = error.response?.data as {detail: string};
        throw new Error(`${detail.detail}`)
    });
};

// Eliminate an account
export const deleteAccount = (token: string, id?: number): Promise<string> => {
    return axi.delete(`/accounts/${id}`, { headers: { 'Authorization': `Bearer ${token}` } })
    .then(response => {
        console.log(response.status);
        return response.statusText;
    })
    .catch((error: AxiosError) => {
        const detail: {detail: string} = error.response?.data as {detail: string};
        throw new Error(`${detail.detail}`)
    });
};